### PR TITLE
Add a new trait to handle post types

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

A refactor to remove LOTS of repeated code and add a cleaner way of handling post type objects in our code.

## Relevant Technical Choices

A trait is used, as an object doesn't make a tonne of sense here, as it does handle state and these are just helper functions. 

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #

### CODE REVIEW

File: includes/Traits/Post_Type.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	 *
+	 * @since 1.7.0
+	 *
+	 * @param string $slug The post type slug.
+	 * @param string $cap Capability name.
+	 *
+	 * @return string|false
+	 */
+	protected function get_post_type_cap_name( $slug, $cap ) {
+		$post_type_obj = get_post_type_object( $slug );
```
Comment (miina) [inline includes/Traits/Post_Type.php]: Looks like here are some linting issues with spaces.

**Review by miina — APPROVED** [review]

File: includes/Traits/Post_Type.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Traits/Post_Type.php]: Fixed

File: includes/Traits/Post_Type.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	 * Get post type capability on the post type slug and name.
+	 *
+	 * @since 1.7.0
+	 *
+	 * @param string $slug The post type slug.
+	 * @param string $cap Capability name.
+	 *
+	 * @return bool
+	 */
+	protected function get_post_type_cap( $slug, $cap ) {
```
Comment (swissspidy) [inline includes/Traits/Post_Type.php:68]: Nit: this not an ideal name. When reading `get_post_type_cap` I don't expect to get a bool flag.

`current_user_can` or `current_user_can_for_post_type` or similar would be clearer.

File: includes/Traits/Post_Type.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	/**
+	 * Get rest controller on the post type slug.
+	 *
+	 * @since 1.7.0
+	 *
+	 * @param string $slug The post type slug.
+	 *
+	 * @return WP_REST_Posts_Controller|WP_REST_Controller
+	 */
+	protected function get_post_type_parent_controller( $slug ) {
```
Comment (swissspidy) [inline includes/Traits/Post_Type.php:137]: `get_post_type_rest_controller` seems more fitting here.

Comment (swissspidy) [reply includes/Traits/Post_Type.php:137]: Similar, `$parent_controller` could be renamed to `$rest_controller`

File: includes/Traits/Post_Type.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	/**
+	 * Get rest base name based on the post type slug.
+	 *
+	 * @since 1.7.0
+	 *
+	 * @param string $slug The post type slug.
+	 *
+	 * @return string Rest base.
+	 */
+	protected function get_post_type_rest_base( $slug ) {
```
Comment (swissspidy) [inline includes/Traits/Post_Type.php:48]: Renaming `$slug` to `$post_type` everywhere would make it more clear that we're dealing with post type slugs and not post permalink slugs.

File: includes/Traits/Post_Type.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+use WP_Post_Type;
+use WP_REST_Controller;
+use WP_REST_Posts_Controller;
+
+/**
+ * Trait Post_Type
+ *
+ * @package Google\Web_Stories\Traits
+ */
+trait Post_Type {
```
Comment (swissspidy) [inline includes/Traits/Post_Type.php:38]: Nit: pros of making this a trait instead of just namespaced functions?

**Comment by spacedmonkey** [comment]
@swissspidy If you don't like naming, just create a PR to fix it. Commenting on a merged PR, doesn't make a lot of sense to me.

File: includes/Traits/Post_Type.php (hunk 5 above)
Comment (spacedmonkey) [reply includes/Traits/Post_Type.php:38]: With was a suite of helper methods. Makes sense to me to group it together in a trait.

**Comment by swissspidy** [comment]
I'm currently going through all the notifications and review requests from the past two weeks and leave feedback where I see fit. Better than not saying anything at all :-) I don't have the bandwidth to create PRs for everything though.

**Comment by spacedmonkey** [comment]
> I don't have the bandwidth to create PRs for everything though.

Okay then, what should be done about this feedback?

**Pull request merged**